Testing Infrastructure

madhur-tandon opened the pull request:
@mdboom Ideally, using `selenium.run` is equivalent to `pyodide.runPython` and thus, the experimental code below should return the `canvas_data` aka the underlying pixel data. However, this stops executing and throws the following error (inside the `get_pixel_data()` function)

The first 2 lines of this function are:
```
canvas = self.get_element('canvas')
ctx = canvas.getContext("2d")
```

The error comes due to the fact that the variable `canvas` is `None` and thus, a `NoneType` object cannot have the method `getContext`.

Surprisingly however, If I run the same code on an Iodide notebook i.e.
```
%%py

import matplotlib.pyplot as plt
import numpy as np

t = np.arange(0.0, 2.0, 0.01)
s = 1 + np.sin(2 * np.pi * t)

plt.plot(t, s, linewidth=1.0, marker=11)
plt.plot(t, t)
plt.grid(True)
plt.show()
canvas_data = plt.gcf().canvas.get_pixel_data()
print(canvas_data.shape)
```

This works! And I can actually see the shape on the console.

Any Ideas as to why `selenium` takes the value of `None` for the variable `canvas`?

madhur-tandon commented:
@mdboom The selenium tests use the URL `f'http://{server_hostname}:{server_port}/test.html'`
This URL doesn't provide an Iodide frontend and thus, the variable `canvas` above is `None` which kinda makes us stuck at this step. 

Initially, I thought that `plt.plot()` shall take some time and thus, I tried to use the `implicit` and `explicit` wait conditions of selenium (https://selenium-python.readthedocs.io/waits.html) to wait for the buttons in the Navigation Toolbar to appear (which kinda confirms that a figure has been plotted). But then, I discovered that selenium doesn't use the Iodide frontend, it just uses the `test.html` file in the `build` directory. Thus, even using `wait` in selenium is futile as those buttons will never appear in the DOM.

Do you have any suggestions / another approach to make the test work?

mdboom commented:
> This URL doesn't provide an Iodide frontend and thus, the variable canvas above is None which kinda makes us stuck at this step.

The `create_root_element` function is supposed to fall back to something that works without Iodide.  It's possible that's broken and needs to be fixed.  In general, we just need to provide a way to add DOM elements to the document -- for Iodide we have to make sure they go to the right place.  Outside of Iodide, they can just be appended directly to `<body>`.

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +28,17 @@ def test_pdf(selenium):
+    selenium.run("plt.grid(True)")
+    selenium.run("plt.show()")
+    canvas_data = selenium.run("plt.gcf().canvas.get_pixel_data()")
+    assert canvas == '1'
```
A couple thoughts:

1. You could use a multi-line string for this (the tests above in this file predate support for that -- maybe you could update those as well while you're here).

2. You probably don't want or need to return the data over to the pytest side here.  You could just do all the comparison work within Pyodide and return a boolean and assert that that boolean is True.

madhur-tandon replied:
About 2nd, If I don't return data to the pytest side i.e. if I use 
`selenium.run("canvas_data = plt.gcf().canvas.get_pixel_data()")`

How will I verify the data of the reference image which will probably be read from a folder. This is because reading the data of the reference image would always be on the `pytest` side. How can the reference data be put inside the `pyodide` side too?

mdboom replied:
You could fetch the data from the pytest side to the pyodide side.  Files in `test/data` are available from the testing webserver.  The reason I suggest doing it this way is that sending the data back to selenium causes a conversion to Javascript and then to Python again that may be hard to deal with.

madhur-tandon commented:
@mdboom I am now able to load both:
- the canvas data
- the reference data
from the `pyodide` side,

However, there are a couple of roadblocks.

1) The canvas data in the pyodide side is a `numpy` array of the shape `(H, W, 4)`. 
2) To load the reference image, I have used the function `pyodide.open_url`. But this function returns an object of type `io.StringIO` as written here: https://github.com/iodide-project/pyodide/blob/master/docs/api_reference.md#pyodideopen_urlurl

Thus, the 2 quantities are in different formats.

While the canvas data can be manipulated i.e. can be reshaped, converted to `bytes`, etc.

I am unable to see what sort of manipulation I can do with the `reference data`. For now, I have tried to use `io.open(ref_data, encoding='rb', buffering = 0)` but this results in a `ValueError: embedded null character`. Moreover, using `io.open` directly on `test/canvas.png` doesn't seem to work.

The questions I want to ask are as follows:
a) is `pyodide.open_url` the only function available that we could use to get the `reference data` on the pyodide side? This is because a PNG file also contains `metadata` which we do not want, we are only concerned with the underlying pixel data. Using `pyodide.open_url` will also give us the metadata.

b) Do you have an alternate approach as to how to convert these 2 quantities, bring them to the same level, so that we can compare them?

PS - Even if I manipulate the `canvas data` to the format of `io.StringIO`, I suspect we still cannot compare it directly with the `io.StringIO` object returned by `pyodide.open_url`. This is because the latter will also contain the metadata as the reference image was written by `_png` function of matplotlib.

madhur-tandon commented:
@mdboom 
I have modified the `create_root_element` function. While this modification allows us to get the `canvas data` in the testing infrastructure since the `canvas` is not `None` as we can find it in the DOM. 

This messes up `pyodide` when used in local testing with the `iodide` frontend. The plot only appears in the `Report` view i.e. when the `Report` button is clicked on the `iodide` frontend. Lastly, while the navigation toolbar appears, the buttons seem to go `inactive`. I guess we would also have to modify the `create_root_element` function. But I am wondering how these 2 things can be achieved simultaneously i.e.
- the `pyodide` build works as it is with the `iodide` frontend.
- the `canvas` tag is not `None` in the testing infrastructure.

madhur-tandon commented:
@mdboom Experimented with a few conversion methods, I think I might need some help with bringing both the data types (`canvas data` and `reference data`) to the same level for comparison.

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +28,29 @@ def test_pdf(selenium):
+    # assert ref_data == 1
+    selenium.run("plt.gcf().canvas.print_png(data)")
+    y = selenium.run("base64.b64encode(data.getvalue()).decode('utf-8')")
+    selenium.run("ref_data = pyodide.open_url('test/canvas.png').getvalue()")
```
`pyodide.open_url` only supports text (Unicode) data, but PNGs are a binary format.  Web browsers can load binary data but only asyncronously, but `open_url` is designed to be a synchronous API.

However, you can use the [fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) to load a blob or `XMLHttpRequest` to load a binary blob.  (There's a pyodide example of this [here](https://github.com/iodide-project/pyodide/blob/020b7f7783faadb257592e31c8cc7875710d83d3/packages/micropip/build/micropip-0.1/micropip.py#L32).  You can then pass that blob into an `io.BytesIO` and pass that to matplotlibs `imread`, which will give you a Numpy array.

madhur-tandon replied:
Looking into this now!

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +28,29 @@ def test_pdf(selenium):
+    # selenium.run("ref_data = pyodide.open_url('test/canvas.png').getvalue()")
+    # ref_data = selenium.run("io.open(ref_data, encoding='rb', buffering = 0)")
+    # assert ref_data == 1
+    selenium.run("plt.gcf().canvas.print_png(data)")
```
You don't want the PNG data here, just the raw pixel data, so just use `get_pixel_data`.

madhur-tandon replied:
Yes, that's what I used earlier, but `pyodide.open_url` was the roadblock there too

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +28,29 @@ def test_pdf(selenium):
+    selenium.run("ref_data = pyodide.open_url('test/canvas.png').getvalue()")
+    z = selenium.run("base64.b64encode(io.BytesIO(ref_data.encode('utf-8')).getvalue()).decode('utf-8')")
+    # x = selenium.run("base64.b64encode(data.getvalue()).decode('ascii') == ref_data")
+    assert y == z
```
I doubt this will work to send the numpy arrays over the wire like this.  I would just compare them in Python code and return the boolean result of `np.all(y == z)` over to the pytest side.

madhur-tandon replied:
Yes, that makes sense. Sorry about that.

madhur-tandon commented on test/test_matplotlib.py:
```diff
@@ -28,3 +28,29 @@ def test_pdf(selenium):
+    selenium.run("plt.plot(t, s, linewidth=1.0, marker=11)")
+    selenium.run("plt.plot(t, t)")
+    selenium.run("plt.grid(True)")
+    selenium.run("plt.show()")
```
Yes, that is on the plan once we have the test working.

madhur-tandon commented on test/test_matplotlib.py:
```diff
@@ -28,3 +28,35 @@ def test_pdf(selenium):
+    ref_data = _get_url_async('test/canvas.png')
+    """)
+    x = selenium.run("ref_data")
+    assert x == 1
```
@mdboom I have tried numerous ways, and I ultimately settled to this. Here, the value of `x` is `nay`. If I do not return `nay`, it's `None`. Thus, it can be inferred that the code never goes to the condition `if req.readyState == 4:` which kinda means a not successful transfer. I suspect that the URL `test/canvas.png` doesn't work except for the function `pyodide.open_url`. This is because when I ran the test, I could see the status code as `500`. Any tips on how to get this right?

madhur-tandon replied:
Alternatively, I decided to use `fetch`, `Request` and `Response`

Usually, the code for `fetch` on JS is used as
```
var myImage = document.createElement('img');
var myRequest = new Request('https://i.ibb.co/3f4yJQS/face4.jpg');

fetch(myRequest)
.then(function(response) {
  return response.blob();
})
.then(function(myBlob) {
  var objectURL = URL.createObjectURL(myBlob);
  myImage.src = objectURL;
});

document.body.appendChild(myImage);
```

For the python side, I tried the following but got stuck since I do not know how to write promises on Python side.
```
from js import Request, fetch, Response
req = Request.new('test/canvas.png')
resp = fetch(req)
# how to convert resp to blob using Response?
```

madhur-tandon replied:
I have also tried the following snippet on an Iodide Notebook - in an attempt to use the `fetch` API from the Python side.

```
%%py

from js import document, Request, fetch, URL
img_tag = document.createElement('img')
req = Request.new('https://i.ibb.co/3f4yJQS/face4.jpg')

def func(response):
    return response.blob()

def func2(blob):
    objURL = URL.createObjectURL(myBlob)
    img_tag.src = objURL

fetch(req).then(func).then(func2)

document.body.appendChild(img_tag)
```

This yields the error `TypeError: 'fetch' called on an object that does not implement interface Window.` The same error comes on the terminal while running the test too.

mdboom replied:
You can pass Python functions to `then`, just like in Javascript.  The only difference is that Python doesn't really have anonymous functions, so you have to explicitly define them.  So the example above converts to:

```
def get_blob(response):
    return response.blob()

def set_object_url(my_blob):
    object_url = URL.createObjectURL(my_blob)
    myImage.src = object_url

fetch(my_request).then(get_blob).then(set_object_url)
```

madhur-tandon commented:
@mdboom We are finally able to load a `blob` object from the `PNG` file. For some reason, wrapping it up in `io.BytesIO` seems to throw the error `BufferError: Can not use as buffer`. Shouldn't this work though?

madhur-tandon commented:
@mdboom I have updated the test, the code should work. However, waiting for the variable `window.result` always leads to a `TimeOutException` even after `timeout` is set to a higher value i.e. 150 in case. Any ideas here as to why it always it always leads to a `timeout`?

mdboom commented:
Perhaps the image comparison is failing, and returning False.  The wait check would never pass in that case.  Perhaps you want to do `return inited is not None` there?

madhur-tandon commented:
@mdboom I have removed the `TimeOutException`. There's one more roadblock due to which our tests our failing and this is regarding the size of the image.

The `ref_data` variable gets its contents from a local PNG and thus has the dimensions `960, 1280, 4`

The `canvas_data` variable gets its contents from the function `get_pixel_data()`. This when run in the `Iodide` front-end yields an image of size `960, 1280, 4` which is good.

However, inside the selenium tests, the variable `canvas_data` has the size of `480, 640, 4` which is half in both width and height.

Due to the mismatch in dimensions, the comparison goes to `False`.

I suspect something is fishy with the `create_root_element` function.

Any pointers on how can we fix this?

mdboom commented:
It sounds like when you run the code when actually displaying to a screen, it's in HiDPI mode.  When running headless in selenium, it's running in regular DPI mode.  You can either force it to be Lo DPI when you generate the reference image, or force it to be HiDPI when running the test by modifying `get_dpi_ratio`.

madhur-tandon commented:
@mdboom Should I manually change the dpi ratio and then generate the PNG file from the Iodide frontend? Or do we have some other option which can be programatically done?

mdboom commented:
What about modifying get_dpi_ratio to detect when it's in a test and return
a hard-coded value?

On Wed, Jun 12, 2019 at 11:25 AM Madhur Tandon <notifications@github.com>
wrote:

> @mdboom <https://github.com/mdboom> Should I manually change the dpi
> ratio and then generate the PNG file from the Iodide frontend? Or do we
> have some other option which can be programatically done?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/iodide-project/pyodide/pull/459?email_source=notifications&email_token=AAAJLFT4NTTLNITROPTEN5LP2EIMZA5CNFSM4HVZWW5KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXQZVUQ#issuecomment-501324498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAJLFWVK2PCVWTJ43IINEDP2EIMZANCNFSM4HVZWW5A>
> .
>


-- 
Michael Droettboom
Data Engineer
Mozilla Corporation

madhur-tandon commented:
@mdboom Two things:

1) about modifying `get_dpi_ratio` if it's in a test, shouldn't we just set it as `1` irrespective of it's in a test or not? (just like in the latest commit I pushed). That way, the dimensions of the PNG image will match the dimensions of SVG and PDF files.

2) I have now used `np.mean(np.abs(canvas_data - ref_data))`.
Interestingly, this is not 0, is this expected?

Moreover, this value is different when the test is tested by `FirefoxDriver` vs `Chromedriver`.

In Firefox, the mean of absolute deviations is `0.8506095377604168`
While for Chrome, it is `2.478726399739583`

mdboom commented:
Does hardcoding the value to 1 make it lower resolution than possible on a retina screen?  The gist here is that you need to make the context in which you generate the reference image and the context in which you test against the reference image to match.  Matching the size of the PNG with the size of the SVG and PDF should be considered a non-goal.  The PNG should match exactly what the user is seeing on their screen.

I would expect a lot of differences between Firefox and Chrome, but not between different runs of the same browser.  This brings up the question of debugging this.  I guess we'll want a way to get the data out so that we can inspect it after Selenium testing.  There is probably a way to modify the mini-webserver that runs the tests to accept a POST request that the test could upload the data to, and the webserver could save the file to the local filesystem.  Maybe there's other ways to get the data out that I haven't thought of.  Sorry to have missed this requirement, but it will be necessary when things don't match well to answer *why*.

SimonBiggs commented:
> This brings up the question of debugging this. I guess we'll want a way to get the data out so that we can inspect it after Selenium testing. There is probably a way to modify the mini-webserver that runs the tests to accept a POST request that the test could upload the data to, and the webserver could save the file to the local filesystem. Maybe there's other ways to get the data out that I haven't thought of. Sorry to have missed this requirement, but it will be necessary when things don't match well to answer _why_.

If it's helpful, over at the Jyve repository there is an example of taking screenshots using headless Firefox during selenium tests on Travis:

https://github.com/deathbeds/jyve/blob/gh-pages/headlessfirefox/kernels/lab/Brython/Notebook_2_win.png

https://github.com/deathbeds/jyve/blob/ee397899d15a83f9edfa085da9e858b7224e101b/tests/acceptance/Kernels.robot#L54

madhur-tandon commented:
@mdboom `get_canvas_data()` inside `conftest.py` gives us the PNG image from the testing web server. We get one image for the `firefox web driver` and the other for the `chrome web driver`. These saved images are then compared to the image data we get from the `get_pixel_data()` function. 

One modification that I did was that the web driver that runs this new test actually uses the PNG that was generated by that web driver itself (aka chrome web driver uses the PNG that was saved by the chrome web driver through the testing web server, and similarly for Firefox)

The mean of absolute deviations is then compared to the threshold `0.1` and if the mean is less than this, the generated PNG images (from the testing webserver) are also removed -- indicating that the tests pass. This also has the advantage that the generated PNGs are made and removed on the fly and thus need not reside in the `test` folder.

Also, if the test fails, the generated PNGs are not removed. Thus, we can always refer to them later and inspect as to why the test failed.

madhur-tandon commented:
@mdboom Also, as to why the mean isn't 0, I guess this has to do with different implementations of a) getting the canvas data using `getImageData()` of the HTML5 Canvas context and b) saving the PNG from the web server using `.toDataURL('image/png')`. These two ways of getting out the data / saving are implicitly different and thus, this might be the reason for the mean not being 0.

madhur-tandon commented:
@mdboom Lastly, due to the modification I added (i.e. chrome web driver uses the PNG that was saved by the chrome web driver through the testing web server, and similarly for Firefox). We now do not see / observe the large deviations that were previously occurring due to a comparison between data from 2 different browsers.

mdboom commented on packages/matplotlib/src/html5_canvas_backend.py:
```diff
@@ -27,6 +27,22 @@ class FigureCanvasHTMLCanvas(FigureCanvasWasm):
+        return root_element
+
+    def get_dpi_ratio(self, context):
+        try:
```
Instead of using exceptions, can you do:

```
if hasattr(window, 'testing'):
    ...
```

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +32,51 @@ def test_pdf(selenium):
+    wait = WebDriverWait(selenium.driver, timeout=70)
+    wait.until(ResultLoaded())
+    res = selenium.run("window.result")
+    assert res is True
```
Combine these two lines to:

```
assert selenium.run("window.result") is True
```

mdboom commented on test/conftest.py:
```diff
@@ -230,6 +230,16 @@ def urls(self):
+        img_script = "return arguments[0].toDataURL('image/png').substring(21)"
+        canvas_base64 = self.driver.execute_script(img_script, canvas_element)
+        canvas_png = base64.b64decode(canvas_base64)
+        with open(r"test/canvas-{0}.png".format(self.browser), 'wb') as f:
```
Hardcoding `test/` here won't work if you run the tests from somewhere other than the root of the source tree.  Can we use `TEST_PATH` instead?

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +32,51 @@ def test_pdf(selenium):
+    wait.until(ResultLoaded())
+    res = selenium.run("window.result")
+    assert res is True
+    os.remove("test/canvas-{0}.png".format(selenium.browser))
```
This should be done in a `finally` clause so the test will clean up after itself even if the test fails.

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +32,51 @@ def test_pdf(selenium):
+
+        req.onreadystatechange = callback
+        req.send(None)
+    """)
```
Lets move this to a helper function in `backend_html5_canvas`, and maybe call it `compare_reference_image`.  There isn't any need to pass in `cb` here -- just call `_png.read_png_int` directly.

madhur-tandon commented:
@mdboom I think this is ready now unless you require some minor changes. I have been able to bring the mean of absolute deviations to 0 🎉. This required re-implementing the function `get_pixel_data` so that it too uses `toDataURL('image/png')` just as `get_canvas_data` in `conftest.py` was using. Thus, the pathways for the `live test data` and the `reference data` are the same.

Lastly, I have also included a check so as to not generate the reference data again and again for each run of the test. The reference data is only generated if it doesn't exist.

This should be ready 💃🏼

mdboom commented:
The code looks really good now.  However, it does look like the test is failing in Circle CI.  The easiest way to debug these things is often to SSH into the Circle CI build.  Let me know on Gitter if you don't have permissions and I can figure out how to give them to you.

madhur-tandon commented:
@mdboom This is finally ready! The `chrome tests` that fail are non-deterministic and are not related to the test introduced by this PR.

mdboom commented on packages/matplotlib/src/html5_canvas_backend.py:
```diff
@@ -48,19 +64,33 @@ def draw(self):
-        return pixel_data
+        img_URL = canvas.toDataURL('image/png')[21:]
+        canvas_base64 = base64.b64decode(img_URL)
+        return _png.read_png_int(io.BytesIO(canvas_base64))
```
Can you add a comment here about why it takes the long way round?  pixels -> encode png -> decode png -> pixels.  (Because not doing it results in different data than the reference image...)

mdboom commented on test/test_matplotlib.py:
```diff
@@ -28,3 +32,40 @@ def test_pdf(selenium):
+    plt.show()
+    """)
+
+    if not os.path.isfile('test/canvas-{0}.png'.format(selenium.browser)):
```
Can you add a comment here: "If we don't have a reference image, write one to disk"

rth commented on test/conftest.py:
```diff
@@ -230,6 +230,17 @@ def urls(self):
+        canvas_png = base64.b64decode(canvas_base64)
+        with open(TEST_PATH /
+                  r"canvas-{0}.png".format(self.browser), 'wb') as f:
+            f.write(canvas_png)
```
Is this going to be used outside of the matplotlib test? If not it might be better to move it to the matplotlib file directly as a function, if possible.

madhur-tandon replied:
By the `matplotlib` file, do you mean the new backend aka the file `html5_canvas_backend.py`? This wouldn't be possible since the `backend` does not have access to `selenium` (which is ideal). OR are you talking about the file `test_matplotlib.py` file? This could be done but would require some extra lines that define the variable `TEST_PATH`. What is preferred?

mdboom replied:
I think @rth's suggestion of moving this to the `test_matplotlib.py` file makes sense, given that it is matplotlib-specific.

rth commented on test/test_matplotlib.py:
```diff
@@ -1,3 +1,7 @@
+import os
+from selenium.webdriver.support.wait import WebDriverWait
```
not that this file was moved to `packages/matplotlib/test_matplotlib.py` -- you will see it if you merge master in.

madhur-tandon replied:
Yes, I noticed, I guess we can do that once the whole `gsoc` branch gets merged with `master` i.e. we let them be separate OR should I actually move it inside the `packages/matplotlib` directory?

mdboom replied:
Soon (doesn't have to be in this PR), we'll want to rebase the gsoc branch on the master branch to "shorten the distance" between the two.

rth commented on test/test_matplotlib.py:
```diff
@@ -28,3 +32,40 @@ def test_pdf(selenium):
+    """)
+
+    if not os.path.isfile('test/canvas-{0}.png'.format(selenium.browser)):
+        selenium.get_canvas_data()
```
Does it take long to run? if not might be good to delete it at the end of the test and always re-create it.

mdboom replied:
We don't want to recreate it every time -- it's just a convenient way to create it the first time.  This is to create a reference image that then subsequent runs (and runs on CI) would use.

rth replied:
Sure, that also works.

mdboom commented:
\o/
